### DISCUSSION TITLE
simple math function "sin()" seems slower than expected under Python implementation of duckdb

### DISCUSSION THREAD

**Discussion by stonebig** [duckdb/duckdb#8699] [General]
hi, 

doing this test  on pytthon duckdb and https://shell.duckdb.org/

```
WITH RECURSIVE
  cnt(x) AS (
     SELECT 1
     UNION ALL
     SELECT x+1 FROM cnt
      where x < 300000 
  )

select sum(sin(x))  from cnt 
```

On my PC and imperfect test method, It takes about:
- 10 seconds on my pc with duckdb-0.8.2.dev3092-cp311-cp311-win_amd64.whl,
- versus 4.8 seconds on the web site shell.duckdb.org v0.8.2-dev1781
- versus 1 seconds on SQLite with math_functions

Is it something expected ?

![image](https://github.com/duckdb/duckdb/assets/4312421/9f8b1eb5-4739-4cac-8dcf-1ee880eac7f8)

**Comment by stonebig** [comment]
```
import duckdb
con = duckdb.connect()

import time
start_time = time.time()
sql="""WITH RECURSIVE
  cnt(x) AS (
     SELECT 1
     UNION ALL
     SELECT x+1 FROM cnt
      where x < 300000 
  )

select sum(sin(x))  from cnt
"""
results = duckdb.sql(sql).fetchall()

print(results)
print("--- %s seconds ---" % (time.time() - start_time))
```

gives

```
[(1.8787588791128815,)]
--- 10.472519874572754 seconds ---
 
```

**Comment by l1t1** [comment]
the bottle neck is cte, not sin()
VS postgresql
```sql
~ $ duckdb
v0.7.1 b00b93f
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D .timer on
D WITH RECURSIVE
>   cnt(x) AS (
>      SELECT 1
>      UNION ALL
>      SELECT x+1 FROM cnt
>       where x < 300000
>   )
>
> select sum(sin(x))  from cnt ;
100% ▕████████████████████████████████████████████████████████████▏
┌────────────────────┐
│    sum(sin(x))     │
│       double       │
├────────────────────┤
│ 1.8787588791128726 │
└────────────────────┘
Run Time (s): real 125.544 user 138.033676 sys 81.679556

D create table cnt as WITH RECURSIVE
>   cnt(x) AS (
>      SELECT 1
>      UNION ALL
>      SELECT x+1 FROM cnt
>       where x < 300000
>   )select * from cnt;
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 178.950 user 155.891914 sys 91.838243
D select sum(sin(x))  from cnt ;
┌──────────────────┐
│   sum(sin(x))    │
│      double      │
├──────────────────┤
│ 1.87875887911291 │
└──────────────────┘
Run Time (s): real 0.014 user 0.024235 sys 0.001034
--postgresql 15.3
mydb=# WITH RECURSIVE
mydb-#   cnt(x) AS (
mydb(#      SELECT 1
mydb(#      UNION ALL
mydb(#      SELECT x+1 FROM cnt
mydb(#       where x < 300000
mydb(#   )
mydb-#
mydb-# select sum(sin(x))  from cnt ;
        sum
--------------------
 1.8787588791128726
(1 row)

Time: 911.808 ms
```

**Comment by l1t1** [comment]
another test on duckdb 0.7.1
```sql
D select sum(sin(x))  from range(1,300001)cnt(x) ;
┌────────────────────┐
│    sum(sin(x))     │
│       double       │
├────────────────────┤
│ 1.8787588791128726 │
└────────────────────┘
Run Time (s): real 0.020 user 0.016199 sys 0.000403
```

**Comment by stonebig** [comment]
quite interesting:
- this is now very very fast :  `select sum(sin(x))  from range(1,300000)cnt(x) ;`
- and this is twice less slower, taking 10 seconds  than before 20 seconds
    `select sum(py_sin(x))  from range(1,30000)cnt(x) ;`

Py_sin being defined as 

for sqlite:
```
pydef py_sin(s) 
    import math as py_math
    return ("%s" %  py_math.sin(s*1));
```

for duckdb:
```
pydef py_sin(s:int) ->float:
    import math as py_math
    return ( py_math.sin(s*1));
```

so:
- **recursive** CTE mecanism seems quite slow in duckdb (maybe per design)
- calling UDF function seems apparently also very slow in that test, compared to SQLite, for strange reason.

**Comment by l1t1** [comment]
it's reasonable that custom defined function is slower than build in one, for context switching

**Comment by stonebig** [comment]
> it's reasonable that custom defined function is slower than build in one, for context switching

Sure, but way slower than SQLite, that has to go through string transformation ,is curious.
( 3 000 loop/second limit for dusckDB, versus 300 000 loop/second for SQLite)


it's of course "corner case" experiments, no real-life use-case,  but to explore what duckDB is good/not good at vs SQLite, in its python wheel form.

**Comment by FrancoisLepoutre** [comment]
Hi,

No willing to get into a technical debate here. I am not enough of a techie to provide great input. But IMHO Duckdb top value is at mangling large sets of data, meaning really large ones. Specifically running calculations, ordering data and downloading and uploading significant data subsets. That's per "columnar" design. 

When it comes to the speed of individual sql calls including recursive cte calls or massive udf usage, I understand that these operations may certainly be more costly than sqlite ones. Hannes has been very cautious on udf-s. And you can understand that a "columnar engine" brings massive power on large operations. But, in some way, there is a cost on lighter ones. In my view it make sense that a "line-based engine" or even more so a record-based datastore would be faster on such stuff.

As far as I am concerned I find this is certainly an acceptable pattern. Should you need speedy record-wise operations that incur a significant calculation cost within the python eco-system, downloading to, say, numpy, running your costly operations via numba, pythran, cython and friends. (and possibly feeding back to our Duckdb) is the way to go.

**Reply by stonebig** [reply]
Exploring  the weakness  points,  their reasons, and the possible remedies.
Is there a link to Hannes comments on udf-s ?

**Comment by l1t1** [comment]
find a faster way to generate data using RECURSIVE CTE
```sql
with RECURSIVE t(lv,x)
as
(select 1 lv, x from (values(0),(1))a(x)
union all
select t.lv+1,s.a+t.x from t ,(values(0),(1<<t.lv))s(a) where t.lv<20)
select * from t where lv=20 order by x;

├───────┴─────────┤
│  1048576 rows   │
│   (40 shown)    │
└─────────────────┘
Run Time (s): real 0.546 user 0.920406 sys 0.046800
```

**Comment by stonebig** [comment]
About slow Python UDF, there was a discussion there: https://github.com/duckdb/duckdb/discussions/8206

**Reply by FrancoisLepoutre** [reply]
Hi stonebig.  Yes you tacked it:-) You may also read https://github.com/duckdb/duckdb/issues/471

Also I am making sure to stay on the python user-side of db engine tech, I found the discussion of UDFs by Hannes and Mark (Mytherin) quite enlightening!
